✂️ Remove the approvals gem ✂️

vaidehijoshi opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR removes the `approvals` gem, which should unblock us in [our journey to upgrading to Rails 6.1](https://github.com/forem/forem/pull/11333#issuecomment-781475076)! 🎉 

This PR:
1. Deletes fixtures in `/spec/support/fixtures/approvals/*`
2. Deletes fixtures in `/spec/fixtures/approvals/*`
3. Ports over any tests that _were_ using `Approvals.verify` to use Rspec's built-in `include` matcher instead.

The only thing I that remains is a handful of tests in `spec/services/feeds/import_spec.rb`, which I'm hoping to get some help with (as I'm a bit unfamiliar with the feeds importer code and wasn't sure as to the best way to port over the tests to _not_ use the Approvals methods)

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/12955

## QA Instructions, Screenshots, Recordings

Make sure the tests pass! 😎 

### UI accessibility concerns?

_Nada!_

## Added tests?

- [ ] Yes
- [x] No, and this is why: _Updated pre-existing tests only_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## Are there any post deployment tasks we need to perform?

Nope!

## What gif best describes this PR or how it makes you feel?

![seal of approval](https://media0.giphy.com/media/13zeE9qQNC5IKk/giphy.gif?cid=5a38a5a2mski6lt7a0wwwfjdk7p0idj4ihixut3lz6zu5ydk&rid=giphy.gif)

vaidehijoshi commented on spec/liquid_tags/codepen_tag_spec.rb:
```diff
@@ -17,10 +17,23 @@ def generate_new_liquid(link)
       Liquid::Template.parse("{% codepen #{link} %}")
     end
 
+    def page_with_codepen_tag
+      <<~HTML
+        <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
+        <html xmlns="http://www.w3.org/1999/xhtml">
+          <body>
+            <iframe height="600" src="https://codepen.io/twhite96/embed/XKqrJX?height=600&amp;default-tab=result&amp;embed-version=2" scrolling="no" frameborder="no" allowtransparency="true" loading="lazy" style="width: 100%;">
+            </iframe>
+          </body>
+        </html>
+      HTML
+    end
```
This is me attempting to move [`spec/support/fixtures/approvals/codepen_liquid_tag.approved.html`](https://github.com/forem/forem/blob/c0ea4e58787a1142ddb952fdbe2709f177158ec0/spec/support/fixtures/approvals/codepen_liquid_tag.approved.html) inline and out of the approvals fixture file.

Is this along the lines that you were thinking, @joshpuetz? (This actually doesn't work for some reason, I haven't investigated deeply yet but if you see what I'm doing wrong, please let me know 😕)

citizen428 replied:
Reasons this is failing:

1. The approvals gem seems to strip/ignore whitespace, this does not.
2. `page_with_codepen_tag` uses `&amp;` but `rendered_codepen_iframe ` includes an actual `&` character.

Here's what I did to get the spec to pass:

```diff
diff --git a/spec/liquid_tags/codepen_tag_spec.rb b/spec/liquid_tags/codepen_tag_spec.rb
index 93967a436..224a38cc3 100644
--- a/spec/liquid_tags/codepen_tag_spec.rb
+++ b/spec/liquid_tags/codepen_tag_spec.rb
@@ -18,11 +18,11 @@ def generate_new_liquid(link)
     end

     def page_with_codepen_tag
-      <<~HTML
+      <<~HTML.gsub(/\s+/, " ")
         <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
         <html xmlns="http://www.w3.org/1999/xhtml">
           <body>
-            <iframe height="600" src="https://codepen.io/twhite96/embed/XKqrJX?height=600&amp;default-tab=result&amp;embed-version=2" scrolling="no" frameborder="no" allowtransparency="true" loading="lazy" style="width: 100%;">
+            <iframe height="600" src="https://codepen.io/twhite96/embed/XKqrJX?height=600&default-tab=result&embed-version=2" scrolling="no" frameborder="no" allowtransparency="true" loading="lazy" style="width: 100%;">
             </iframe>
           </body>
         </html>
@@ -32,8 +32,9 @@ def page_with_codepen_tag
     it "accepts codepen link" do
       liquid = generate_new_liquid(codepen_link)
       rendered_codepen_iframe = liquid.render
+      expected = rendered_codepen_iframe.gsub(/\s+/, " ")
       # Approvals.verify(rendered_codepen_iframe, name: "codepen_liquid_tag", format: :html)
-      expect(page_with_codepen_tag).to include(rendered_codepen_iframe)
+      expect(page_with_codepen_tag).to include(expected)
     end
```

But, I'm not sure if I see any real benefit to using this style of approval for the liquid tags at all. We're using `include` as a matcher, so nothing else in `page_with_codepen_tag` really matters. We might as well just check that `liquid.render` produces the expected output.

joshpuetz replied:
Yeah, that's more or less what I was thinking @vaidehijoshi. And I agree with @citizen428: many of these refactors will lead to a point where it's easier to just do a build in rspec matcher!

vaidehijoshi commented:
Okay, I think that removing this gem might be _slightly_ more involved than we initially thought: there are actually 27 tests that use fixtures from the approvals gem, not 10. 🙈 It wasn't obvious to me when I picked this up, but once I started running the specs I realized that the approvals fixtures live in two places:

1. `/spec/support/fixtures/approvals/` as well as
2. `/spec/fixtures/approvals`

So far, I've just skipped the specs that I've identified as relying on approvals fixtures in some way. I took a stab at porting over a single test but it doesn't pass at the moment.

vaidehijoshi commented on spec/services/feeds/import_spec.rb:
```diff
@@ -12,7 +12,7 @@
     end
 
-    it "fetch only articles from a feed_url", vcr: { cassette_name: "feeds_import" } do
+    xit "fetch only articles from a feed_url", vcr: { cassette_name: "feeds_import" } do
```
I could use help with the tests in this file! I can see that the feeds importer service is returning a number, but I can't quite figure out _what_ exactly we're testing in these specs that I've skipped here.

Maybe @rhymes or @maestromac have some suggestions on how to convert these away from using the `approvals` gem's `verify` (which won't work anymore), into something that is actually useful here? 🤔

rhymes replied:
I think we're just testing that `num_articles` is 21, that's it :D To make sure it's always importing the same amount of `<item>`s in the RSS feed (you could either hardcode the number or, probably better, XML-parse the file and check that `num_articles` corresponds to the number of `<item>` tags, maybe?)

This was borrowed by an older `RssReader` spec

vaidehijoshi commented on spec/system/user_uses_the_editor_spec.rb:
```diff
@@ -47,7 +47,13 @@ def fill_markdown_with(content)
       article_body.gsub!(%r{"https://res\.cloudinary\.com/.{1,}"}, "cloudinary_link")
 
-      Approvals.verify(article_body, name: "user_preview_article_body", format: :html)
+      # TODO: Convert this to an E2E test?
```
I think we probably want to convert this into a E2E test, but I didn't want to increase the scope of this too much, so I added some tests using the `user_preview_article_body.approved.html` fixture file for now and left a note to come back to it in the future.

rhymes replied:
yeah, good call!

vaidehijoshi commented on spec/system/user_uses_the_editor_spec.rb:
```diff
@@ -58,7 +64,13 @@ def fill_markdown_with(content)
       article_body.gsub!(%r{"https://res\.cloudinary\.com/.{1,}"}, "cloudinary_link")
 
-      Approvals.verify(article_body, name: "user_preview_article_body", format: :html)
+      # TODO: Convert this to an E2E test?
```
I think we probably want to convert this into a E2E test, but I didn't want to increase the scope of this too much, so I added some tests using the `user_preview_article_body.approved.html` fixture file for now and left a note to come back to it in the future.

rhymes commented on .dockerignore:
```diff
@@ -20,7 +20,6 @@ vendor/bundle
 /tmp
 .DS_Store
 .swp
-.approvals
```
should we keep it? This one once this is merged people won't have an unknown `.approvals` file they need to manually delete and might accidentally re-commit or re-submit into a PR

vaidehijoshi replied:
That's fair! I re-added this back into `.dockerignore` in the last commit :)

rhymes commented on .gitignore:
```diff
@@ -14,7 +14,6 @@ vendor/bundle
 /tmp
 .DS_Store
 .swp
-.approvals
```
should we keep it? This one once this is merged people won't have an unknown `.approvals` file they need to manually delete and might accidentally re-commit or re-submit into a PR

vaidehijoshi replied:
That's fair! I re-added this back into `.gitignore` in the last commit :)

rhymes commented on Gemfile:
```diff
@@ -159,7 +159,6 @@ group :development, :test do
 end
 
 group :test do
-  gem "approvals", "~> 0.0" # A library to make it easier to do golden-master style testing in Ruby
```
LOL at `0.0`

rhymes commented on spec/views/email_subscriptions/unsubscribe.html.erb_spec.rb:
```diff
@@ -4,6 +4,6 @@
     assign(:email_type, "#{SiteConfig.community_name} digest emails")
     render
-    Approvals.verify(rendered, name: "email_subscriptions/unsubscribe", format: :html)
+    expect(rendered).to include("You have been unsubscribed from DEV(local) digest emails. 😔")
```
should we avoid hardcoding `DEV(local)`?

vaidehijoshi replied:
Good call! I'll use `SiteConfig.community_name` here instead 👍🏽

rhymes reviewed (commented):
Thanks for taking this on! I see a couple of more specs that are using `approvals`: `spec/liquid_tags/tweet_tag_spec.rb` and `spec/models/github_issue_spec.rb`

vaidehijoshi commented on spec/liquid_tags/tweet_tag_spec.rb:
```diff
@@ -23,11 +23,14 @@ def generate_tweet_liquid_tag(id)
   end
 
-  xit "render properly", :vcr do
+  it "render properly", :vcr do
```
This was a skipped test previously, but now that I've "un-`approvals`"ed it, I think we can unskip it :) 

It seems to consistently pass for me, but if it becomes flaky again or if we need to skip it again down the road, I figured we can cross that bridge when we come to it.

rhymes replied:
agreed! left a non blocking suggestion

```suggestion
  it "renders properly", :vcr do
```

vaidehijoshi commented on spec/models/github_issue_spec.rb:
```diff
@@ -88,10 +88,12 @@
   end
 
-  xit "saves HTML in .processed_html" do
+  it "saves HTML in .processed_html" do
```
This was a skipped test previously, but now that I've "un-`approvals`"ed it, I think we can unskip it :) 

It seems to consistently pass for me, but if it becomes flaky again or if we need to skip it again down the road, I figured we can cross that bridge when we come to it.

vaidehijoshi commented on spec/services/feeds/import_spec.rb:
```diff
@@ -12,10 +12,11 @@
       end
     end
 
+    # TODO: We could probably improve these tests by parsing against the items in the feed rather than hardcoding
```
I was weary of going down a rabbit hole and completely re-writing these tests by parsing the VCR file, so I converted them to a baseline Rspec version, but left a comment here as to how they could be improved!

vaidehijoshi commented on spec/services/feeds/import_spec.rb:
```diff
@@ -87,17 +86,17 @@
     end
 
     context "with an explicit set of users", vcr: { cassette_name: "feeds_import" } do
+      # TODO: We could probably improve these tests by parsing against the items in the feed rather than hardcoding
```
I was weary of going down a rabbit hole and completely re-writing these tests by parsing the VCR file, so I converted them to a baseline Rspec version, but left a comment here as to how they could be improved!

vaidehijoshi commented:
> I see a couple of more specs that are using `approvals`: `spec/liquid_tags/tweet_tag_spec.rb` and `spec/models/github_issue_spec.rb`

@rhymes Oh I totally missed those!! Thank you for catching that. I've updated them to remove `approvals`...and while I was at it, I noticed that they were skipped, but now that I've converted them back to regular old Rspec, I think we can unskip them 😎

vaidehijoshi commented:
I think I've addressed all comments and fixed up the 2 `Approvals` specs that I missed previously! This should be ready for re-review now 🦄 (cc @joshpuetz, @rhymes, @maestromac)

rhymes reviewed (approved):
LGTM!

maestromac reviewed (approved):
LGTM!

vaidehijoshi commented on spec/liquid_tags/instagram_tag_spec.rb:
```diff
@@ -13,8 +13,14 @@ def generate_instagram_tag(id)
-      Approvals.verify(rendered_instagram, name: "instagram_liquid_tag", format: :html)
+
+      # rubocop:disable Style/StringLiterals
+      expect(liquid.render).to include('<iframe')
+      expect(liquid.render).to include('id="instagram-liquid-tag"')
+      expect(liquid.render).to include("https://www.instagram.com/p/#{valid_id}/embed/captioned")
+      expect(liquid.render).to include('src="https://platform.instagram.com/en_US/embeds.js"')
+      expect(liquid.render).to include('<div class="instagram-position">')
```
Oh, _great_ call!! I love it, going to update with this wherever I can 👍🏽
